liquid/validation.go

```diff
@@ -0,0 +1,316 @@
+//   - Each resource must report usage exactly for those AZs that its declared topology requires:
+//     For FlatResourceTopology, only AvailabilityZoneAny is allowed.
+//     For other topologies, all AZs in req.AllAZs must be present (and possibly AvailabilityZoneUnknown, but no others).
+//   - All resources declared with HasQuota = true must report quota (and no others).
+//   - Each resource reporting quota must report it in the way that its declared topology requires:
+//     For AZSeparatedResourceTopology, quota must be reported only on the AZ level, and only for real AZs (not for AvailabilityZoneUnknown).
+//     For all other topologies, quota must be reported only on the resource level.
+//   - All rates declared in info.Rates must be present (and no others).
+//   - Each rate must report usage exactly for those AZs that its declared topology requires:
+//     For FlatRateTopology, only AvailabilityZoneAny is allowed.
+//     For other topologies, all AZs in req.AllAZs must be present (and possibly AvailabilityZoneUnknown, but no others).
+//   - All metrics families declared in info.UsageMetricFamilies must be present (and no others).
+//   - The number of labels on each metric must match the declared label set.
+//
+// Additional validations may be added in the future.
+func ValidateUsageReport(report ServiceUsageReport, req ServiceUsageRequest, info ServiceInfo) error {
+	errs := validateUsageReportImpl(report, req, info)
+	if len(errs) > 0 {
+		// NOTE: Errors get joined with "; " instead of ", " because some errors contain commas themselves.
+		return fmt.Errorf("received ServiceUsageReport is invalid: %s", errs.Join("; "))
+	}
+	return nil
+}
+
+// This is the function that the unit tests call. An ErrorSet is easier to compare against fixtures than the final stringified error.
+func validateUsageReportImpl(report ServiceUsageReport, req ServiceUsageRequest, info ServiceInfo) (errs errorset.ErrorSet) {
+	if report.InfoVersion != info.Version {
+		errs.Addf("received ServiceUsageReport is invalid: expected .InfoVersion = %d, but got %d", info.Version, report.InfoVersion)
+		// assume that all other errors would be aftereffects of the version mismatch, and skip finding them
+		return errs
+	}
+
+	// validate metrics
+	errs.Append(validateMetrics(report.Metrics, info.UsageMetricFamilies, ".UsageMetricFamilies"))
+
+	// validate resource reports
+	for resName, resInfo := range info.Resources {
+		if resInfo.HasQuota && !hasKey(report.Resources, resName) {
+			errs.Addf("missing value for .Resources[%q] (resource was declared with HasQuota = true)", resName)
+		}
+	}
+	for resName, res := range report.Resources {
+		resInfo, exists := info.Resources[resName]
+		if !exists {
+			errs.Addf("unexpected value for .Resources[%q] (resource was not declared)", resName)
+			continue
+		}
+		if !resInfo.HasQuota {
+			errs.Addf("unexpected value for .Resources[%q] (resource was declared with HasQuota = false)", resName)
+			continue
+		}
```

Comment:
The checks in 165-167 and 175-177 imply that a ResourceReport should only be present if the resource is declared with `HasQuota = true`. This is incorrect: We also need reports for resources that only report usage.